File: Algorithm_I_study_plan/rotting_oranges.py
from typing import List

class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        width = len(grid)
        height = len(grid[0])
        to_visit_count = width * height
        visited_count = 0
        steps_count = 0
        
        queue = []
        
        # do to_visit -= 1 for every 0 and visited += 1 for every 2
        for r in range(width):
            for c in range(height):
                if grid[r][c] == 2:
                    queue.append((r, c))
                    visited_count += 1
                elif grid[r][c] == 0:
                    to_visit_count -= 1
                    
        
        if visited_count == to_visit_count:
                return steps_count
        
                    
        while queue:
            l = len(queue)
            for _ in range(l):
                r, c = queue.pop(0)
                for i, j in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                    new_r = r + i
                    new_c = c + j
                    
                    if 0 <= new_r < width and 0 <= new_c < height and grid[new_r][new_c] == 1:
                        grid[new_r][new_c] = 2
                        queue.append((new_r, new_c))
                        visited_count += 1
            
            steps_count += 1
            
            if visited_count == to_visit_count:
                return steps_count
            
        return -1


# slightly cleaner solution
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        minutes = 0
        to_rot_count = 0
        queue = []
        visited = set()
        
        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c] == 2:
                    queue.append((r, c))
                    visited.add((r, c))
                elif grid[r][c] == 1:
                    to_rot_count += 1
        
        # bfs
        l = len(queue)
        while queue:
            
            if to_rot_count == 0:
                return minutes
            
            minutes += 1
            l = len(queue)
            while l:
                r, c = queue.pop(0)

                for x, y in [(1, 0), (0, 1), (-1, 0), (0, -1)]:
                    new_r = r + y
                    new_c = c + x

                    if 0 <= new_r < len(grid) and \
                       0 <= new_c < len(grid[0]) and (new_r, new_c) not in visited \
                       and grid[new_r][new_c] == 1:
                        grid[new_r][new_c] = 2
                        queue.append((new_r, new_c))
                        to_rot_count -= 1
                        visited.add((new_r, new_c))
                l -= 1
        
        if to_rot_count:
            return -1
        return minutes
 

File: NeetCode_Roadmap/rotting_oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int rem_count = 0;
        int minutes = -1;
        queue<pair<int, int>> q;
        for (int i = 0; i < grid.size(); ++i)
        {
            for (int j = 0; j < grid[0].size(); ++j)
            {
                if (grid[i][j] == 2)
                    q.push(make_pair(i, j));
                else if (grid[i][j] == 1)
                    ++rem_count;
            }
        }

        if (rem_count == 0) return 0;

        while (!q.empty())
        {
            int level_size = q.size();
            for (int k = 0; k < level_size; ++k)
            {
                auto [i, j] = q.front();
                q.pop();
                for (auto [di, dj] : {pair(0, 1), pair(1, 0), pair(0, -1), pair(-1, 0)})
                {
                    int new_i = i + di;
                    int new_j = j + dj;
                    if (0 <= new_i && new_i < grid.size() && 0 <= new_j && new_j < grid[0].size()
                        && grid[new_i][new_j] == 1)
                    {
                        grid[new_i][new_j] = 2;
                        q.push(make_pair(new_i, new_j));
                        --rem_count;
                    }
                }
            }
            ++minutes;
        }
        if (rem_count) return -1;
        return minutes;
    }
};
